### PR TITLE
(proof-new) Add builtin proof checker

### PR DESCRIPTION
**Pull request by ajreynol**
This adds the proof checker for TheoryBuiltin, which handles the core proof rules (SCOPE and ASSUME) and all proof rules corresponding to generic operations on Node objects.  This includes proof rules for rewriting and substitution, which are added in this PR.

### CODE REVIEW

File: src/theory/builtin/proof_checker.h [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+ public:
+  BuiltinProofRuleChecker() {}
+  ~BuiltinProofRuleChecker() {}
+  /**
+   * Apply rewrite on n (in witness form). This encapsulates the exact behavior
+   * of a REWRITE step in a proof. Rewriting is performed on the Skolem form of
+   * n.
+   *
+   * @param n The node (in witness form) to rewrite,
+   * @param id The identifier of the rewriter.
```
Comment (HanielB) [inline src/theory/builtin/proof_checker.h:75]: Update and describe default value.

File: src/theory/builtin/proof_checker.h [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+   * n.
+   *
+   * @param n The node (in witness form) to substitute,
+   * @param exp The (set of) equalities (in witness form) corresponding to the
+   * substitution
+   * @return The substituted form of n.
+   */
+  static Node applySubstitution(Node n,
+                                Node exp,
+                                MethodId ids = MethodId::SB_DEFAULT);
```
Comment (HanielB) [inline src/theory/builtin/proof_checker.h:91]: Describe in the doc

File: src/theory/builtin/proof_checker.h [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+                                const std::vector<Node>& exp,
+                                MethodId ids = MethodId::SB_DEFAULT);
+  /** Apply substitution + rewriting
+   *
+   * Combines the above two steps.
+   *
+   * @param n The node (in witness form) to substitute and rewrite,
+   * @param exp The (set of) equalities (in witness form) corresponding to the
+   * substitution
+   * @param id The identifier of the rewriter.
```
Comment (HanielB) [inline src/theory/builtin/proof_checker.h:102]: Update

File: src/theory/builtin/proof_checker.h [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+  Node checkInternal(PfRule id,
+                     const std::vector<Node>& children,
+                     const std::vector<Node>& args) override;
+  /** get method ids */
+  bool getMethodIds(const std::vector<Node>& args,
+                    MethodId& ids,
+                    MethodId& idr,
+                    size_t index);
+  /**
+   * Apply rewrite (on Skolem form). id is the identifier of the rewriter.
```
Comment (HanielB) [inline src/theory/builtin/proof_checker.h:126]: Update

File: src/theory/builtin/proof_checker.h [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+   */
+  static Node applyRewriteExternal(Node n, MethodId idr = MethodId::RW_REWRITE);
+  /**
+   * Apply substitution for n (on Skolem form), where exp is an equality
+   * (or set of equalities) in Witness form. Returns the result of
+   * n * { exp[0] -> exp[1] } in Skolem form.
+   */
+  static Node applySubstitutionExternal(Node n, Node exp, MethodId ids);
+  static Node applySubstitutionExternal(Node n,
```
Comment (HanielB) [inline src/theory/builtin/proof_checker.h:135]: add doc

File: src/theory/builtin/proof_checker.h [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+  /**
+   * Apply rewrite (on Skolem form). id is the identifier of the rewriter.
+   */
+  static Node applyRewriteExternal(Node n, MethodId idr = MethodId::RW_REWRITE);
+  /**
+   * Apply substitution for n (on Skolem form), where exp is an equality
+   * (or set of equalities) in Witness form. Returns the result of
+   * n * { exp[0] -> exp[1] } in Skolem form.
+   */
+  static Node applySubstitutionExternal(Node n, Node exp, MethodId ids);
```
Comment (HanielB) [inline src/theory/builtin/proof_checker.h:134]: Doc `ids`

File: src/theory/builtin/proof_checker.cpp [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+  pc->registerChecker(PfRule::MACRO_SR_EQ_INTRO, this);
+  pc->registerChecker(PfRule::MACRO_SR_PRED_INTRO, this);
+  pc->registerChecker(PfRule::MACRO_SR_PRED_ELIM, this);
+  pc->registerChecker(PfRule::MACRO_SR_PRED_TRANSFORM, this);
+}
+
+Node BuiltinProofRuleChecker::applyRewrite(Node n, MethodId idr)
+{
+  Node nk = ProofSkolemCache::getSkolemForm(n);
+  Node nkr = applyRewriteExternal(n, idr);
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/builtin/proof_checker.cpp:67]: yes good catch

File: src/theory/builtin/proof_checker.h [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+  RW_REWRITE,
+  // identity
+  RW_IDENTITY,
+  //---------------------------- Substitutions
+  // (= x y) is interpreted as x -> y, using Node::substitute
+  SB_DEFAULT,
+  // P, (not P) are interpreted as P -> true, P -> false using Node::substitute
+  SB_LITERAL,
+  // P is interpreted as P -> true using Node::substitute
+  SB_FORMULA,
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/builtin/proof_checker.h:52]: I think this is the name I prefer. The second means "view this formula as a literal" (e.g. an assignment to true/false), the latter means "view this formula as a formula" (e.g. it is interpreted as true).

**Pull request merged**